### PR TITLE
PCHR-1245 : Fetch location label instead of value in vacancies extension

### PR DESCRIPTION
**Pull request by omarabuhussein**
## Problem 

On Vacancy ( HRRecruitment ) extension pages the location value was used instead of it's label , Here is are some screenshots :

selection_114

selection_113

## Solution

I changed location field to a pseudo constant and then used CRM_Core_PseudoConstant::getLabel() to retrieve the label from the location value whenever it is used through the extension . 

I also did a minor changes in CRM_HRRecruitment_Page_WidgetJs::vacancyInfo() to use CRM_Utils_Array::value() method to retrieve array values via keys to remove notices appear when these array keys are missingt.

selection_117

selection_116

### CODE REVIEW

File: hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php [hunk 1]
```diff
@@ -243,6 +243,9 @@ static function &fields()
           'title' => ts('Location') ,
           'maxlength' => 254,
           'size' => CRM_Utils_Type::HUGE,
+          'pseudoconstant' => array(
+            'optionGroupName' => 'hrjc_location',
+          ),
```
Comment (davialexandre) [inline hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php]: Please also add this to the schema xml file

File: hrrecruitment/CRM/HRRecruitment/BAO/HRVacancy.php [hunk 2]
```diff
@@ -194,10 +194,15 @@ static function getVacanciesByStatus() {
...
           $vacancy['start_date'] = !empty($vacancy['start_date']) ? CRM_Utils_Date::customFormat($vacancy['start_date'], '%b %E, %Y') : NULL;
           $vacancy['end_date'] = !empty($vacancy['end_date']) ? CRM_Utils_Date::customFormat($vacancy['end_date'], '%b %E, %Y') : NULL;
+
+          $locationValue = CRM_Utils_Array::value('location', $vacancy);
+          if ($locationValue !== FALSE) {
+            $locationValue = CRM_Core_PseudoConstant::getLabel('CRM_HRRecruitment_BAO_HRVacancy', 'location', $locationValue);
+          }
+
           $vacancyEntry[$vacancy['status_id']]['vacancies'][$id] = array(
-            'date' => CRM_Utils_Date::customFormat($vacancy['start_date'], '%b %E, %Y') . ' - ' . CRM_Utils_Date::customFormat($vacancy['end_date'], '%b %E, %Y'),
```
Comment (davialexandre) [inline hrrecruitment/CRM/HRRecruitment/BAO/HRVacancy.php]: why the date was removed?

Comment (omarabuhussein) [inline hrrecruitment/CRM/HRRecruitment/BAO/HRVacancy.php]: the key is duplicated twice

File: hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php (hunk 1 above)
Comment (omarabuhussein) [inline hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php]: oh my bad, I should start using you script to generate DAO from XML files.

File: hrrecruitment/CRM/HRRecruitment/Page/WidgetJs.php [hunk 3]
```diff
@@ -117,7 +117,13 @@ public static function vacancyListDisplay() {
       $row[$vacancyVal['id']]['id'] = $vacancyVal['id'];
       $row[$vacancyVal['id']]['position'] = CRM_Utils_Array::value('position', $vacancyVal);
       $row[$vacancyVal['id']]['positionLink'] = "{$base_url}/civicrm/vacancy/info?id={$vacancyVal['id']}";
-      $row[$vacancyVal['id']]['location'] = CRM_Utils_Array::value('location', $vacancyVal);
+
+      $locationValue = CRM_Utils_Array::value('location', $vacancyVal);
+      if ($locationValue !== FALSE) {
+        $locationValue = CRM_Core_PseudoConstant::getLabel('CRM_HRRecruitment_BAO_HRVacancy', 'location', $locationValue);
+      }
```
Comment (davialexandre) [inline hrrecruitment/CRM/HRRecruitment/Page/WidgetJs.php]: These 4 lines of code are duplicated in 4 places (twice in this file). Can we have it moved to a helper method somewhere else? I honestly don't know where would be the best place to put it (maybe in CRM_HRRecruitment_BAO_HRVacancy?), but anywhere will be better than having it repeated in a lot of places

File: hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php [hunk 4]
```diff
@@ -165,67 +125,64 @@ class CRM_HRRecruitment_DAO_HRVacancy extends CRM_Core_DAO
...
-        new CRM_Core_Reference_Basic(self::getTableName() , 'created_id', 'civicrm_contact', 'id') ,
-      );
+  static function getReferenceColumns() {
+    if (!isset(Civi::$statics[__CLASS__]['links'])) {
+      Civi::$statics[__CLASS__]['links'] = static ::createReferenceColumns(__CLASS__);
+      Civi::$statics[__CLASS__]['links'][] = new CRM_Core_Reference_Basic(self::getTableName() , 'created_id', 'civicrm_contact', 'id');
+      CRM_Core_DAO_AllCoreTables::invoke(__CLASS__, 'links_callback', Civi::$statics[__CLASS__]['links']);
     }
-    return self::$_links;
+    return Civi::$statics[__CLASS__]['links'];
```
Comment (davialexandre) [inline hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php]: You expect Civi to improve it's code base and then they come up with things like this (╯°□°）╯︵ ┻━┻

**Comment by davialexandre** [comment]
Overall this looks good, but I have found a problem. When trying to run the tests locally, I got this error:

`PHP Fatal error:  Call to undefined method CRM_Core_DAO_AllCoreTables::invoke() in /vagrant/hr16/sites/all/modules/civicrm/tools/extensions/civihr/hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php on line 250`

It appears you're using the 4.7.11-rc of CiviCRM, right? This `invoke` method was introduced as part of the refactoring Tim did to the DAOs (https://github.com/civicrm/civicrm-core/commit/740dd877cde870c7a80d7e0d1b143e5dee8fb37f#diff-cc0edc0d2e3fb4e89c17bfda27d597afR271). This scares me a bit since it means we won't be able to use a DAO generated on 4.7.11 on previous versions. Why should anyone care about semver, right?!
(╯°□°）╯︵ ┻━┻

Anyway, can you please regenerate the DAO on anything prior to 4.7.11?

File: hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php (hunk 4 above)
Comment (omarabuhussein) [inline hrrecruitment/CRM/HRRecruitment/DAO/HRVacancy.php]: :smile:

**Comment by omarabuhussein** [comment]
Yeah my current instance created with civicrm master branch ( a little bit behind the master actually ) ... thanks for spotting this out .

**Comment by davialexandre** [comment]
OK, so now I was to run the tests, but it seems your changes have introduced a regression. One of the existing tests, which is passing on staging, is failing with the changes on your branch:

captura de tela 2016-09-06 as 08 05 28

Can you please check that?

**Comment by davialexandre** [comment]
All tests passing now. Approved.

**Pull request merged**